### PR TITLE
Refactor Annotations

### PR DESCRIPTION
**Pull request by jackkoenig**
This isn't quite done, but I think it's close and ready for feedback from interested parties. @dt27182 @sdtwigg @stevenmburns @aayupov @seldridge @edcote

This does a lot so pardon the large diff:
* Rename old `Annotation` to `LegacyAnnotation` and deprecate
    * ~**I intend to remove `LegacyAnnotation` in this PR, see discussion below**~
* Change `Annotation` to a trait that can be extended
* Change all existing Annotations to new API
    * This is the bulk of the changes
* Add JsonProtocol for automatic serialization and deserialization of Annotations (so long as they're case classes)
* Replace `AnnotationMap` with `AnnotationSeq` and implicit conversion to and from `Seq`
* Remove Driver.compile
    * (because it accepts `AnnotationMap` and if I'm breaking it might as well remove, open to keeping)

## Remove rather than deprecate old annotations

In this PR I reclaim `firrtl.annotations.Annotation` for the new API which breaks any code that relies on the old fields. Even if we decided to keep `LegacyAnnotation` under the class `Annotation` there is a bigger problem that makes me think the old API should just be removed either way.

**Problem:** How to handle old style annotations (ie. YAML) with refactored internal annotations?
We could just not update any of the existing annotations to the new API until a later date when we finally do remove the old annotations. I don't like this because the whole point of this refactoring is to remove all of the boilerplate. That being said it would make it possible to deprecate before breaking.

Otherwise, assuming we do update the existing annotations to the new API, I can't think of a good way to handle old-style annotations (eg. from a YAML file) that target new annotations. We'd have to keep around all of the boilerplate anyway and ensure that we do not use the autogenerated unapply methods for case class Annotations (this could actually be really tricky since custom unapply functions probably match the signature that the autogenerated one would have). I'm also worried about implementing this being error-prone and possibly leading to annotations being silently ignored.

In my opinion we should just take the pain now and ditch the old annotations implementation. Thoughts?

### I changed my mind on this.

Instead of removing all YAML and LegacyAnnotation stuff, I've added some migration code to convert from old-style to new-style but print a nasty warning. User code that uses LegacyAnnotations is intended to still work albeit with a warning.

## Notes on Migration

Basically, users should make case classes that extend Annotation (or better yet, one of the subtraits):

```scala
/** If an Annotation does not target any [[Named]] thing in the circuit, then all updates just
  * return the Annotation itself
  */
trait NoTargetAnnotation extends Annotation {
  def update(renames: RenameMap) = Seq(this)
}

/** An Annotation that targets a single [[Named]] thing */
trait SingleTargetAnnotation[T <: Named] extends Annotation {
  val target: T

  /** Create another instance of this Annotation */
  def duplicate(n: T): Annotation
}

trait SingleStringAnnotation extends NoTargetAnnotation {
  def value: String
}

// TODO: Add MultiTargetAnnotation
```

Example on how to migrate an annotation that is just checked to see if it exists (like an option to a transform):
```scala
// Old style, note the unapply is just returning true or false!
object NoDCEAnnotation {
  val marker = "noDCE!"
  val transform = classOf[DeadCodeElimination]
  def apply(): Annotation = Annotation(CircuitTopName, transform, marker)
  def unapply(a: Annotation): Boolean = a match {
    case Annotation(_, targetXform, value) if targetXform == transform && value == marker => true
    case _ => false
  }
}

// New style, you can see if this exists with merely annos.contains(NoDCEAnnotation)
case object NoDCEAnnotation extends NoTargetAnnotation
```

Example on how to migrate an annotation that targets something in the Firrtl graph
```scala
// Old version, DontTouch is just pointing to a single ComponentName
object DontTouchAnnotation {
  private val marker = "DONTtouch!"
  def apply(target: ComponentName): Annotation = Annotation(target, classOf[Transform], marker)

  def unapply(a: Annotation): Option[ComponentName] = a match {
    case Annotation(component: ComponentName, _, value) if value == marker => Some(component)
    case _ => None
  }

// New version, parameterized by ComponentName and just needs to implement update
case class DontTouchAnnotation(target: ComponentName) extends SingleTargetAnnotation[ComponentName] {
  def duplicate(n: ComponentName) = this.copy(n)
}
```

## TODO

- [x] ~Remove LegacyAnnotation and YAML utils~ Add migration path and tests
- [ ] Update Chisel and other repos to use this API
- [ ] Find a way to allow Chisel to target Firrtl annotations (make a subclass of named that Chisel's HasId can implement?)
- [ ] Think about the consequences of Emitter Annotations now being serializable (the whole circuit will serialize by default)
- [x] Add Annotation for invoking custom transforms (remove that logic from Chisel)
- [ ] Replace AnnotationSeq and RenameMap in CircuitState with new Metadata type that holds both
- [ ] Remove RunTransformAnnotation, provide alternative API in Chisel

### CODE REVIEW

**Comment by jackkoenig** [comment]
This required refactoring some transforms

@chick in particular, I made a lot of changes to BlackBoxSourceHelper, mind taking a look at that in particular (and the whole PR 😄)

**Comment by jackkoenig** [comment]
Todo:

- [ ] Remove LegacyAnnotation and YAML utils
- [ ] Update Chisel and other repos to use this API
- [ ] Find a way to allow Chisel to target Firrtl annotations (make a subclass of named that Chisel's HasId can implement?)
- [ ] Think about the consequences of Emitter Annotations now being serializable (the whole circuit will serialize by default).

**Comment by edcote** [comment]
Thanks for the info.  Unfortunate timing for me as I'm right now committing some transforms using the old API into my code base.  There is a good case for this API change and, so long as the new API is properly unit tested, the proposed solution looks great.

**Comment by dt27182** [comment]
Looks good to me. We don't have any legacy annotations, so we don't have a code refactoring problem.

About the JSON serialization, would you consider using proto serialization instead? Could be a good way to prototype using proto as a serialization format for the rest of firrtl.

**Comment by edwardcwang** [comment]
Or better yet, why not both? Or perhaps make the backend serialization agnostic and extensible instead?

We could probably just do that in a future pull request though.

**Comment by jackkoenig** [comment]
@dt27182 
The reason I chose JSON over protobuf was that I want the ability to define methods on the Annotation subclasses. There's also the notion of including arbitrary Scala types (which you could provide a serialization/deserialization Json format for) in your annotations. This seems like it might be hard to do with protobuf which is inherently designed to be language-agnostic, but I'll see if there's a way to accomplish both goals with protobuf.

@edwardcwang 
Multiple serialization formats might work well so long as all of the data can be automagically serialized and deserialized. If the writer of an annotation needs to provide custom [de]serialization formats then I don't see how this could work well. That being said, json4s is kind of a standard AST and there might be converters to many other formats.

File: src/test/scala/firrtlTests/AnnotationTests.scala [hunk 1]
```diff
@@ -136,19 +128,23 @@ class AnnotationTests extends AnnotationSpec with Matchers {
...
+      def execute(state: CircuitState) = state.copy(annotations = Seq())
     }
+    val transform = new DeletingTransform
+    val tname = transform.name
     val inlineAnn = InlineAnnotation(CircuitName("Top"))
-    val result = compiler.compile(CircuitState(parse(input), ChirrtlForm, getAMap(inlineAnn)), Seq(new DeletingTransform))
-    result.annotations.get.annotations.head should matchPattern {
-      case DeletedAnnotation(x, inlineAnn) =>
+    val result = compiler.compile(CircuitState(parse(input), ChirrtlForm, Seq(inlineAnn)), Seq(transform))
+    println(result.annotations.head)
```
Comment (azidar) [inline src/test/scala/firrtlTests/AnnotationTests.scala]: Remove println

**Comment by azidar** [comment]
Closes #609

**Comment by jackkoenig** [comment]
Fixes #475

File: src/main/scala/firrtl/Emitter.scala [hunk 2]
```diff
@@ -140,19 +95,15 @@ sealed abstract class FirrtlEmitter(form: CircuitForm) extends Transform with Em
...
 
   override def execute(state: CircuitState): CircuitState = {
-    val newAnnos = getMyAnnotations(state).flatMap {
-      case EmitCircuitAnnotation() =>
+    val newAnnos = state.annotations.flatMap {
+      case EmitCircuitAnnotation(_) =>
         Seq(EmittedFirrtlCircuitAnnotation.apply(
               EmittedFirrtlCircuit(state.circuit.main, state.circuit.serialize)))
-      case EmitAllModulesAnnotation() =>
+      case EmitAllModulesAnnotation(_) =>
```
[the remark this answers is not in the document]
Comment (jackkoenig) [reply src/main/scala/firrtl/Emitter.scala]: I haven't addressed this yet, but it was ignored in the past and this PR merely preserved that behavior, however it probably *should* be used now, it's a way of having multiple emitters run and targeting the right emitter, we might also just use a class hierarchy of emissions annotations instead

**Pull request merged**